cpp/src/arrow/compute/kernels/vector_rank.cc

```diff
@@ -25,6 +25,120 @@ namespace {
 // ----------------------------------------------------------------------
 // Rank implementation
 
+template <typename T>
+Result<Datum> CreateRankings(ExecContext* ctx, const NullPartitionResult& sorted,
+                             const NullPlacement null_placement,
+                             const RankOptions::Tiebreaker tiebreaker,
+                             std::function<T(int64_t)> value_selector) {
```

Comment:
Ideally, the `std::function` overhead would be removed by the compiler after inlining, but perhaps it's better to avoid it upfront (including in debug builds) by passing this as a template parameter instead?
```c++
template <typename ValueSelector, typename T = std::invoke_result_t<ValueSelector, int64_t>>
Result<Datum> CreateRankings(ExecContext* ctx, const NullPartitionResult& sorted,
                             const NullPlacement null_placement,
                             const RankOptions::Tiebreaker tiebreaker,
                             ValueSelector&& value_selector) {
```